enhancements/update/targeted-update-edge-blocking.md

```diff
@@ -0,0 +1,400 @@
+For each edge declared by a release image (primary metadata), the graph-builder will check the blocker cache for matching blocks.
+Edges with no matching blocks are unconditionally recommended, and will be included in `edges`.
+Edges with matching blocks are conditionally recommended, and will be included in `conditionalEdges`.
+
+### Cluster-version operator support for the enhanced schema
+
+The cluster-version operator will learn to parse [`conditionalEdges`](#enhanced-cincinnati-json-representation) into [`conditionalUpdates`](#enhanced-clusterversion-representation).
+`edges` will continue to go straight into `availableUpdates`.
+The operator will log an error if the same target is included in both `edges` and `conditionalEdges`, but will prefer the `conditionalEdges` entry in that case.
+FIXME: maybe restructure `conditionalEdges` so duplicate entries are impossible?
+
+Additionally, the operator will continually re-evaluate the blocking conditionals in `conditionalUpdates` and update `conditionalUpdates[].conditions` accordingly.
+The timing of the evaluation and fresheness are largely internal details.
+The operator can periodically poll, or use edge triggered watches, combinations of watching and polling, etc.
+FIXME: [Do we need to report on recommendation freshness?](#reporting-recommendation-freshness)
+
+To perform the PromQL request, the operator will FIXME: details about connecting to the local Thanos/Prometheus.
+
+FIXME: [`platform` or other properties that are easier to evaluate than PromQL?](#non-promql-filters)
+
+If there are issues evaluating a conditional update, the operator will set the `Unknown` status on the `Recommended` condition.
+The operator will grow a new `warning`-level `CannotEvaluateConditionalUpdates` alert that fires if `lastTransitionTime` for any `Recommended=Unknown` condition is over an hour old.
+
+Any `conditionalUpdates` with `Recommended=True` will have its release inserted into `availableUpdates`.
+
+Both `availableUpdates` and `conditionalUpdates` should be sorted in decreasing SemVer order for stability, to avoid unnecessary status churn.
+
+### Update client support for the enhanced schema
+
+[The web-console][web-console] and [`oc`][oc] both consume ClusterVersion to present clients with a list of available updates.
+With this enhancement, they will both be extended to consume [`conditionalUpdates`](#enhanced-clusterversion-representation).
+When listing recommended updates, clients will list the contents of `availableUpdates`.
+When listing all supported updates, clients will additionally include entries from `conditionalUpdates` with `Recommended!=True`, and include the `reason` and `message` from the `Recommended` condition alongside the supported-but-not-recommended updates.
+Clients may optionally provision for reporting additional condition types, in case new types are added in the future.
+
+### User Stories
```

Comment:
i'd like to see a little more about the admin experience here.  Some of this may be my own lack of experience w/ the upgrade UX, but working an example of what an admin sees when the following potential updates for their cluster exist:

1) an unconditioned edge
2) a conditioned edge that their cluster qualifies for
3) a conditioned edge that their cluster does not qualify for
4) a conditioned edge that could not be evaluated

how/where is that information presented, and what are the implications if they want to choose each of the 4.  (e.g. when is a force required)

Related to this, could we deliberately use this feature to force an admin to decide if an upgrade edge was appropriate for them or not?  (the condition would never evaluate to "true", but would direct them to documentation for how they can decide if they want to consume the update or not....and then if they did decide to do so, how would they do so(force?))

Reply:
> ...and then if they did decide to do so, how would they do so(force?)...

There is a common misconception that ClusterVersion's `force` and `oc adm upgrade`'s `--force` are related to an update edge's being recommended.  In fact, they are orthogonal. 
 openshift/api#885 has been in flight without much interest so far to remove some misleading godocs.

So the only guards around supported/recommended updates today are in ClusterVersion clients like `oc` and the web console.  I haven't been all that prescriptive about their handling if the new cases here.  And I don't see a need to distinguish between your (1) and (2).  The lines slightly above your comment here are my attempt at (3) and (4), as two separate conditions for "conditional and not known to be recommended to you".  Would walking through some mock `oc` flows be helpful to ground the abstract discussion in examples?

Reply:
> There is a common misconception that ClusterVersion's force and oc adm upgrade's --force are related to an update edge's being recommended. 

let me try again:

and then if they did decide to do so, how would they do so(--allow-explicit-upgrade?))

>  And I don't see a need to distinguish between your (1) and (2). 

as a risk averse admin, i want to know if you're recommending this edge to everyone or not.  I don't trust you to tell me my cluster isn't going to be affected.  (maybe my prom data is wrong, maybe you didn't root cause the problem correctly yet, maybe your detection logic is wrong.... i don't want to be the guinea pig)

>  The lines slightly above your comment here are my attempt at (3) and (4), as two separate conditions for "conditional and not known to be recommended to you".

i don't see case (4) called out there ("we failed to evaluate the condition")

>  Would walking through some mock oc flows be helpful to ground the abstract discussion in examples?

sample output would be great

Reply:
> i don't see case (4) called out there ("we failed to evaluate the condition")

That would be `Recommended=Unknown`, with a reason and message.  Other points are good, and I'll add an `oc` workthrough to try and hit all of them.